include/envoy/network/listen_socket.h

```diff
@@ -34,5 +34,65 @@ class ListenSocket {
 typedef std::unique_ptr<ListenSocket> ListenSocketPtr;
 typedef std::shared_ptr<ListenSocket> ListenSocketSharedPtr;
 
+/**
+ * An abstract accepted socket.
+ */
+class AcceptedSocket {
+public:
+  virtual ~AcceptedSocket() {}
+
+  /**
+   * @return the address that the socket was received at, or an original destination address if
+   * applicable.
+   */
+  virtual Address::InstanceConstSharedPtr localAddress() const PURE;
+
+  /**
+   * Reset the original destination address of the socket to a different address than the one
+   * the socket was accepted at.
+   */
+  virtual void resetLocalAddress(const Address::InstanceConstSharedPtr& local_address) PURE;
+
+  /**
+   * @return true if the local address has been reset.
+   */
+  virtual bool localAddressReset() const PURE;
+
+  /**
+   * @return the address that the socket was received from, or an original source address if
+   * applicable.
+   */
+  virtual Address::InstanceConstSharedPtr remoteAddress() const PURE;
+
+  /**
+   * Set the original source address of the socket
+   */
+  virtual void resetRemoteAddress(const Address::InstanceConstSharedPtr& remote_address) PURE;
+
+  /**
+   * @return fd the accepted socket's file descriptor.
+   */
+  virtual int fd() const PURE;
```

Comment:
Per @ggreenway, can we hide the fd from the filters and potentially just expose peek operations for now? I feel like fd can be an implementation detail and passed around as needed into the real connection when it is created. In general, we should heavily limit what filters can do on a non-const accepted socket. Can we try to limit this interface as much as possible, at least from the filter perspective? This might require a base interface just for use by filters.

Reply:
Note that PROXY protocol consumes bytes from the wire, so peek operation alone wouldn't work (but it's also needed).

Reply:
Right, good point. We can have a read, etc. method also if needed.

Reply:
Socket option access is needed by the original dst filter, so that should be added in as well. On the other hand I don't know what the transport socket PR will bring. Maybe keep these listener filters and the interface as they are now and reconsider after the transport socket work is in?

Reply:
I agree that leaving `fd()` in this PR should be fine (unless @mattklein123 and/or @ggreenway have strong feelings against it), just please add `TODO` and hide the socket-level access in another PR (perhaps even after all other changes are merged?).

Reply:
I think it's fine to leave `fd()` in this PR, but per @PiotrSikora please add a TODO. We should also remove the fd transfer/move function from the public interface. That should be an internal implementation detail. Then in a followup we can remove `fd()` and replace with the few accessor functions we need.

Reply:
Added a TODO.